### PR TITLE
Support new resource: alicloud_ess_notification

### PR DESCRIPTION
**Pull request by TalentNing**
=== RUN   TestAccAlicloudEssNotification_basic
--- PASS: TestAccAlicloudEssNotification_basic (73.40s)
PASS
ok      github.com/terraform-providers/terraform-provider-alicloud/alicloud     73.461s

### CODE REVIEW

File: alicloud/resource_alicloud_ess_notification.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	if len(notificationTypeList) > 0 {
+		for _, v := range notificationTypeList {
+			notificationTypes = append(notificationTypes, v.(string))
+		}
+	}
+	return notificationTypes
+}
+
+func parseEssNofiticationId(id string) (scalingGroupId string, notificationArn string, err error) {
+	parts, err := ParseResourceId(id, 6)
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ess_notification.go]: i think should use SplitN(id,":",2)

File: website/docs/r/ess_notification.html.markdown [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+page_title: "Alicloud: alicloud_ess_notification"
+sidebar_current: "docs-alicloud-resource-ess-notification"
+description: |-
+  Provides a ESS notification resource.
+---
+
+# alicloud\_ess\_notification
+
+Provides a ESS notification resource. More about Ess notification, see [Autoscaling Notification](https://www.alibabacloud.com/help/doc-detail/71114.htm).
+
```
Comment (YY-tomorrow) [inline website/docs/r/ess_notification.html.markdown]: add a note Available in 1.55.0+

File: website/docs/r/ess_notification.html.markdown [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+## Attribute Reference
+
+The following attributes are exported:
+
+* `id` - The ID of notification resource, which is composed of 'scaling_group_id' and 'notification_arn' in the format of 'scaling_group_id:notification_arn'.
+* `scaling_group_id` - The ID of the Auto Scaling group.
+* `notification_arn` - The Alibaba Cloud Resource Name (ARN) for the notification object.
+* `notification_types` - The notification types of Auto Scaling events and resource changes.
```
Comment (YY-tomorrow) [inline website/docs/r/ess_notification.html.markdown]: only left id

File: alicloud/resource_alicloud_ess_notification.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		if IsExceptedErrors(err, []string{InvalidNotificationNotFound, InvalidScalingGroupIdNotFound}) {
+			return nil
+		}
+		return WrapErrorf(err, DefaultErrorMsg, d.Id(), request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+	addDebug(request.GetActionName(), raw)
+	return WrapError(essService.WaitForEssNotification(d.Id(), Deleted, DefaultTimeout))
+}
+
+func buildCreateAlicloudEssNotificationRequest(d *schema.ResourceData) *ess.CreateNotificationConfigurationRequest {
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ess_notification.go]: this not need a new method

File: alicloud/resource_alicloud_ess_notification.go (hunk 1 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_notification.go]: fixed

File: website/docs/r/ess_notification.html.markdown (hunk 2 above)
Comment (TalentNing) [reply website/docs/r/ess_notification.html.markdown]: I have added a NOTE.

File: website/docs/r/ess_notification.html.markdown (hunk 3 above)
Comment (TalentNing) [reply website/docs/r/ess_notification.html.markdown]: done

File: alicloud/resource_alicloud_ess_notification.go (hunk 4 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_notification.go]: removed

File: alicloud/resource_alicloud_ess_notification.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+		if IsExceptedErrors(err, []string{InvalidNotificationNotFound, InvalidScalingGroupIdNotFound}) {
+			return nil
+		}
+		return WrapErrorf(err, DefaultErrorMsg, d.Id(), request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+	addDebug(request.GetActionName(), raw)
+	return WrapError(essService.WaitForEssNotification(d.Id(), Deleted, DefaultTimeout))
+}
+
+func notificationTypeSet2List(notificationTypeSet *schema.Set) []string {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ess_notification.go]: I think the method `notificationTypeSet2List ` is redundant.

File: alicloud/resource_alicloud_ess_notification.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+	notificationTypeList := notificationTypeSet.List()
+	if len(notificationTypeList) > 0 {
+		for _, v := range notificationTypeList {
+			notificationTypes = append(notificationTypes, v.(string))
+		}
+	}
+	return notificationTypes
+}
+
+func parseEssNofiticationId(id string) (scalingGroupId string, notificationArn string, err error) {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ess_notification.go]: I think the method `parseEssNofiticationId` is redundant and just use `strings.SplitN` is ok.

File: website/docs/r/ess_notification.html.markdown [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+-> **NOTE:** Supported notification types: 'AUTOSCALING:SCALE_OUT_SUCCESS', 'AUTOSCALING:SCALE_IN_SUCCESS', 'AUTOSCALING:SCALE_OUT_ERROR', 'AUTOSCALING:SCALE_IN_ERROR', 'AUTOSCALING:SCALE_REJECT', 'AUTOSCALING:SCALE_OUT_START', 'AUTOSCALING:SCALE_IN_START', 'AUTOSCALING:SCHEDULE_TASK_EXPIRING'.
+
+
+
+## Attribute Reference
+
+The following attributes are exported:
+
+* `id` - The ID of notification resource, which is composed of 'scaling_group_id' and 'notification_arn' in the format of 'scaling_group_id:notification_arn'.
```
Comment (xiaozhu36) [inline website/docs/r/ess_notification.html.markdown]: "in the format of `<scaling_group_id>:<notification_arn>`"

File: website/docs/r/ess_notification.html.markdown [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `scaling_group_id` - (Required, ForceNew) The ID of the Auto Scaling group.
+* `notification_arn` - (Required, ForceNew) The Alibaba Cloud Resource Name (ARN) for the notification object.
+* `notification_types` - (Optional) The notification types of Auto Scaling events and resource changes. 
+
+-> **NOTE:** The format of `notification_arn` is acs:ess:{region}:{account-id}:{resource-relative-id}. Valid values for `resource-relative-id`: 'cloudmonitor', 'queue/{queuename}', 'topic/topicname'.
```
Comment (xiaozhu36) [inline website/docs/r/ess_notification.html.markdown]: 'cloudmonitor', 'queue/<queue name>', 'topic/<topic name>'.

File: alicloud/resource_alicloud_ess_notification.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAlicloudEssNotificationCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+
+	request := ess.CreateCreateNotificationConfigurationRequest()
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ess_notification.go]: request.RegionId = client.RegionId

File: alicloud/resource_alicloud_ess_notification.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+		}
+	}
+
+	raw, err := client.WithEssClient(func(essClient *ess.Client) (interface{}, error) {
+		return essClient.CreateNotificationConfiguration(request)
+	})
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, "alicloud_ess_notification", request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+	addDebug(request.GetActionName(), raw)
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ess_notification.go]: addDebug(request.GetActionName(), raw, request.RpcRequest, request)

File: alicloud/resource_alicloud_ess_notification.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+			request.NotificationType = &notificationTypes
+		}
+	}
+	raw, err := client.WithEssClient(func(essClient *ess.Client) (interface{}, error) {
+		return essClient.ModifyNotificationConfiguration(request)
+	})
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, d.Id(), request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+	addDebug(request.GetActionName(), raw)
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ess_notification.go]: still

File: alicloud/resource_alicloud_ess_notification.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+	raw, err := client.WithEssClient(func(essClient *ess.Client) (interface{}, error) {
+		return essClient.DeleteNotificationConfiguration(request)
+	})
+	if err != nil {
+		if IsExceptedErrors(err, []string{InvalidNotificationNotFound, InvalidScalingGroupIdNotFound}) {
+			return nil
+		}
+		return WrapErrorf(err, DefaultErrorMsg, d.Id(), request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+	addDebug(request.GetActionName(), raw)
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ess_notification.go]: still

File: alicloud/resource_alicloud_ess_notification.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+	return resourceAlicloudEssNotificationRead(d, meta)
+}
+
+func resourceAlicloudEssNotificationDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	essService := EssService{client}
+	request := ess.CreateDeleteNotificationConfigurationRequest()
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ess_notification.go]: request.RegionId

File: alicloud/resource_alicloud_ess_notification.go (hunk 5 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_notification.go]: removed

File: alicloud/resource_alicloud_ess_notification.go (hunk 6 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_notification.go]: removed

File: website/docs/r/ess_notification.html.markdown (hunk 7 above)
Comment (TalentNing) [reply website/docs/r/ess_notification.html.markdown]: 修改了

File: website/docs/r/ess_notification.html.markdown (hunk 8 above)
Comment (TalentNing) [reply website/docs/r/ess_notification.html.markdown]: 修改了

File: alicloud/resource_alicloud_ess_notification.go (hunk 9 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_notification.go]: 增加了

File: alicloud/resource_alicloud_ess_notification.go (hunk 10 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_notification.go]: 增加了

File: alicloud/resource_alicloud_ess_notification.go (hunk 11 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_notification.go]: 修改了

File: alicloud/resource_alicloud_ess_notification.go (hunk 12 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_notification.go]: 修改了

File: alicloud/resource_alicloud_ess_notification.go (hunk 13 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_notification.go]: 修改了

**Pull request merged**